### ISSUE TITLE
Undef behavior in plasticity kernel?

### ISSUE THREAD

**Issue by krenzland** [SeisSol/SeisSol#960]
**Describe the bug**
I get the following warnings when compiling (current master with slight modifications NOT in plasticity, but with an ALIGNMENT=128bytes):
```
/Users/lukaskrenz/src/SeisSol/src/Kernels/Plasticity.cpp:198:25: warning: iteration 336 invokes undefined behavior [-Waggressive-loop-optimizations]
  198 |         dudt_pstrain[q] = factor * (prev_degreesOfFreedom[q] - degreesOfFreedom[q]);
      |         ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/lukaskrenz/src/SeisSol/src/Kernels/Plasticity.cpp:173:30: note: within this loop
  173 |       for (unsigned q = 0; q < 6 * NUMBER_OF_ALIGNED_BASIS_FUNCTIONS; ++q) {
      |                            ~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/lukaskrenz/src/SeisSol/src/Kernels/Plasticity.cpp:210:22: warning: iteration 56 invokes undefined behavior [-Waggressive-loop-optimizations]
  210 |         QEtaModal[q] = pstrain[6 * NUMBER_OF_ALIGNED_BASIS_FUNCTIONS + q];
      |         ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/lukaskrenz/src/SeisSol/src/Kernels/Plasticity.cpp:209:30: note: within this loop
  209 |       for (unsigned q = 0; q < NUMBER_OF_ALIGNED_BASIS_FUNCTIONS; ++q) {
      |                            ~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/lukaskrenz/src/SeisSol/src/Kernels/Plasticity.cpp:238:73: warning: iteration 56 invokes undefined behavior [-Waggressive-loop-optimizations]
  238 |         pstrain[6 * NUMBER_OF_ALIGNED_BASIS_FUNCTIONS + q] = QEtaModal[q];
      |                                                              ~~~~~~~~~~~^
/Users/lukaskrenz/src/SeisSol/src/Kernels/Plasticity.cpp:237:30: note: within this loop
  237 |       for (unsigned q = 0; q < NUMBER_OF_ALIGNED_BASIS_FUNCTIONS; ++q) {
      |                            ~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/lukaskrenz/src/SeisSol/src/Kernels/Plasticity.cpp:210:22: warning: 'void* __builtin_memcpy(void*, const void*, long unsigned int)' writing 512 bytes into a region of size 448 overflows the destination [-Wstringop-overflow=]
  210 |         QEtaModal[q] = pstrain[6 * NUMBER_OF_ALIGNED_BASIS_FUNCTIONS + q];
      |         ~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/lukaskrenz/src/SeisSol/src/Kernels/Plasticity.cpp:71:10: note: destination object 'QEtaModal' of size 448
   71 |     real QEtaModal[tensor::QEtaModal::size()] __attribute__((aligned(ALIGNMENT)));
      |          ^~~~~~~~~
/Users/lukaskrenz/src/SeisSol/src/Kernels/Plasticity.cpp:238:60: warning: 'void* __builtin_memcpy(void*, const void*, long unsigned int)' reading 512 bytes from a region of size 448 [-Wstringop-overread]
  238 |         pstrain[6 * NUMBER_OF_ALIGNED_BASIS_FUNCTIONS + q] = QEtaModal[q];
      |         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~
/Users/lukaskrenz/src/SeisSol/src/Kernels/Plasticity.cpp:71:10: note: source object 'QEtaModal' of size 448
   71 |     real QEtaModal[tensor::QEtaModal::size()] __attribute__((aligned(ALIGNMENT)));
```

**Expected behavior**
No warnings.

**To Reproduce**
1. Which version do you use? 1579dfa17603dedca54cfc324ac89390f35a727b + modified ALIGNMENT
2. Which build settings do you use? Which compiler version do you use? gcc-13, Apple M2, rel with deb info
3. On which machine does your problem occur? If on a cluster: Which modules are loaded? MacBook Pro M2
4. Provide parameter/material files. n/a

**Screenshots/Console output**
n/a

**Additional context**
Currently adding support for M1/M2 + MacOS. I'll check the issue later, this is just as a note for me :)

**Comment by davschneller** [comment]
That begs the question... Is the `NUMBER_OF_ALIGNED_BASIS_FUNCTIONS` still correct here? Or what would happen if that would be replaced by `NUMBER_OF_BASIS_FUNCTIONS`?

**Comment by krenzland** [comment]
Fixed by #963
